src/librustdoc/json/mod.rs

```diff
@@ -1,47 +1,237 @@
         &mut self,
-        _item: &clean::Item,
-        _item_name: &str,
-        _cache: &Cache,
+        item: &clean::Item,
+        _module_name: &str,
+        cache: &Cache,
     ) -> Result<(), Error> {
-        unimplemented!()
+        use clean::types::ItemEnum::*;
+        if let ModuleItem(m) = &item.inner {
+            for item in &m.items {
+                match &item.inner {
+                    // These don't have names so they don't get added to the output by default
+                    ImportItem(_) => self.item(item.clone(), cache).unwrap(),
+                    ExternCrateItem(_, _) => self.item(item.clone(), cache).unwrap(),
+                    ImplItem(i) => {
+                        i.items.iter().for_each(|i| self.item(i.clone(), cache).unwrap())
+                    }
+                    _ => {}
+                }
+            }
+        }
+        self.item(item.clone(), cache).unwrap();
+        Ok(())
     }
 
     fn mod_item_out(&mut self, _item_name: &str) -> Result<(), Error> {
-        unimplemented!()
+        Ok(())
     }
 
-    fn after_krate(&mut self, _krate: &clean::Crate, _cache: &Cache) -> Result<(), Error> {
-        unimplemented!()
+    fn after_krate(&mut self, krate: &clean::Crate, cache: &Cache) -> Result<(), Error> {
+        debug!("Done with crate");
+        let mut index = (*self.index).clone().into_inner();
+        index.extend(self.get_trait_items(cache));
+        let output = types::Crate {
+            root: types::Id(String::from("0:0")),
+            version: krate.version.clone(),
+            includes_private: cache.document_private,
+            index,
+            paths: cache
+                .paths
+                .clone()
+                .into_iter()
+                .chain(cache.external_paths.clone().into_iter())
+                .map(|(k, (path, kind))| {
+                    (
+                        k.into(),
+                        types::ItemSummary { crate_id: k.krate.as_u32(), path, kind: kind.into() },
+                    )
+                })
+                .collect(),
+            external_crates: cache
+                .extern_locations
+                .iter()
+                .map(|(k, v)| {
+                    (
+                        k.as_u32(),
+                        types::ExternalCrate {
+                            name: v.0.clone(),
+                            html_root_url: match &v.2 {
+                                ExternalLocation::Remote(s) => Some(s.clone()),
+                                _ => None,
+                            },
+                        },
+                    )
+                })
+                .collect(),
+            format_version: 1,
+        };
+        let mut p = self.out_path.clone();
+        p.push(output.index.get(&output.root).unwrap().name.clone().unwrap());
+        p.set_extension("json");
+        serde_json::ser::to_writer_pretty(&File::create(p).unwrap(), &output).unwrap();
```

Comment:
This panics if `p` is not a directory.
```
rustdoc +stage1 --output-format json x.rs -o /dev/stdout
thread 'rustc' panicked at 'called `Result::unwrap()` on an `Err` value: Os { code: 20, kind: Other, message: "Not a directory" }', src/librustdoc/json/mod.rs:230:60
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace

error: internal compiler error: unexpected panic
```
Is there any reason to output to `$p/crate.json` instead of `$p` directly? I guess you don't know if it came the user or from another part of rustdoc ...

Reply:
Ok, how about this idea: if `p` is a directory, create a file in it (which handles the default `doc/` case), otherwise write to it directly.

Reply:
The current error for HTML output is
```
error: couldn't generate documentation: File exists (os error 17)
  |
  = note: failed to create or modify "/dev/null"
```
so I'm just going to go with that in the first draft.